File: script/game.lua
local DRAW = require"draw"
local WORLD = require"world"
local CAMERA = require"camera"
local STATE = require"state"
local dbg = require"dbg"

local game = STATE.extend{
  world = nil,
  px = 0,
  py = 0,
  pauseNext = false,
}

function game.new(m)
  local o = setmetatable({}, {__index = game})
  o:init(m)
  prettyPrint(o)
  return o
end

function game:init(m)
  DRAW.init("#spritesheet")
  self.world = WORLD.load(m)
  self.cam = CAMERA.new(0, 0, VIEW_WIDTH, VIEW_HEIGHT)
  self.cam:limit(self.world.width, self.world.height)
  self.cam:margin(10, 9)
end

function game:update()
  if self.pauseNext == true then
    JSPROG.pause = true
    self.pauseNext = false
  end
end

function game:draw(dt, ctx)
  ctx.clearRect(0, 0, 320, 240)
  ctx.fillStyle = "black"
  ctx.fillRect(0, 0, 320, 240)

  self.cam:drawWorld(ctx, self.world)
  self.cam:drawSprite(ctx, 26, self.px, self.py)
  self.cam:follow(self.px, self.py)
end

function game:keydown(k)
  local dx = 0
  local dy = 0 
  if k == "w" then dy = -1
  elseif k == "a" then dx = -1
  elseif k == "s" then dy = 1
  elseif k == "d" then dx = 1
  elseif k == "p" then JSPROG.pause = not JSPROG.pause
  elseif k == "z" then dbg.b()
  elseif k == "." then 
    JSPROG.pause = false
    self.pauseNext = true
  elseif k == "0" then 
    reloadAll()
  end

  if self.world:test(self.px + dx, self.py + dy) == false then
    self.px = self.px + dx
    self.py = self.py + dy
  end
end

return game

File: src/game.js
import {Camera, PutTile, DrawMap} from "./render";
import {FromLayer} from "./map2d";

export class Game {
  constructor(name, mapData) {
    this.mapName = name; 
    this.acc = 0;
    this.xx = 0;
    this.yy = 0;
    this.ev = new Set();
    this.tiles = FromLayer(mapData, "map");
    this.col = FromLayer(mapData, "collision");
    this.cam = new Camera(27, 20, 10, 8);
    this.cam.limit(this.tiles.w, this.tiles.h);
  }

  update(dt) {
    this.acc += dt;
    // make a shallow copy
    const cc = this.ev;
    this.ev = new Set();
    let dx = 0;
    let dy = 0;
    for (let e of cc) {
      switch (e) {
        case "R":
          dx += 1;
          break;
        case "L":
          dx -= 1;
          break;
        case "U":
          dy -= 1;
          break;
        case "D":
          dy += 1;
          break;
      }
    }

    if (this.col.get(this.xx + dx, this.yy + dy) == -1) {
      this.xx += dx; 
      this.yy += dy;
    }

    this.cam.follow(this.xx, this.yy);
  }

  draw(_, ctx) {
    ctx.fillStyle = "black";
    ctx.fillRect(0,0,320,240);
    DrawMap(ctx, this.tiles, this.cam);
    PutTile(ctx, 26, (this.xx - this.cam.xx) * 12, (this.yy - this.cam.yy)*12);
  }

  keydown(e) {
    if (e == "a") {
      this.ev.add("L");
    } else if (e == "d") {
      this.ev.add("R");
    } else if (e == "s") {
      this.ev.add("D");
    } else if (e == "w") {
      this.ev.add("U");
    } else if (e == "e") {
      this.ev.add("X");
    }
  }
}

